jobs: produce SHOW JOBS without reading legacy info rows

dt opened the pull request:
Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138104)
<!-- Reviewable:end -->

dt commented:
Here's a first peek at the read query to power crdb_internal.jobs without reading job_info. That said, this is default off -- and maybe doesn't merge -- until we get auth checks which I think will require rebasing on #138102, but I wanted to look at the query here in isolation, then talk about the auth checks once everything else is settled (maybe as a separate PR even).

msbutler commented on pkg/sql/crdb_internal.go:
```diff
@@ -1409,6 +1432,134 @@ func makeJobsTableRows(
+			state,
+			status,
+			created,
+			created, // deprecated "started" field.
```
make the started field `tree.DNull`?

dt replied:
I donno, as long as we have a column called "started" you can see, I think it is more confusing to say it hasn't started -- null -- than to show you the time when it was created and nodes started being able to adopt and run it. Eventually I'd rather remove this column from every surface, but for now it seemed least-disruptive to just populate it with the creation time?

msbutler replied:
sounds good.

msbutler commented on pkg/sql/crdb_internal.go:
```diff
@@ -1409,6 +1432,134 @@ func makeJobsTableRows(
+				tree.NewDString(payloadType.String()),
+				tree.NewDString(job.Description),
+				tree.NewDString(job.Username.Normalized()),
+				tree.NewDString(string(jobs.StatusRunning)),
```
the legacy version marks these session jobs as `StatusPending`.

dt replied:
Ah, so _that's_ where we use this status!

dt commented on pkg/sql/crdb_internal.go:
Done.

msbutler commented on pkg/sql/crdb_internal.go:
```diff
@@ -1409,6 +1432,134 @@ func makeJobsTableRows(
+			id, typStr, desc, ownerStr, state, status, created, finished, modified, fraction, resolved, errorMsg, instanceID =
+				r[0], r[1], r[2], r[3], r[4], r[5], r[6], r[7], r[8], r[9], r[10], r[11], r[12]
+
+			owner := username.MakeSQLUsernameFromPreNormalizedString(string(tree.MustBeDString(ownerStr)))
```
Planning to rebase on https://github.com/cockroachdb/cockroach/pull/138102?

msbutler reviewed (approved):
LGTM! You could also make this setting metamorphic once this is rebased on https://github.com/cockroachdb/cockroach/pull/138102 ? At the very least, you could give this a spin on backup_restore_roundtrip.go roachtests.

dt committed:
jobs: remove TestShowJobsWithError

This isn't a very useful test: it is checking how SHOW JOBS
handles invalid bytes in legacy_progress, which isn't something
we actually need to account for (and changes when we don't read
this row at all soon).

Release note: none.
Epic: none.

dt committed:
jobs: produce SHOW JOBS without reading legacy info row

Release note (sql change): SHOW JOBS is now based on a new mechanism for storing information about the progress and status of running jobs.
Epic: none.

dt commented:
Okay, I've rebased this and made it on by default. Let's see what the tests think now.

dt commented:
Only the last two commits here are outstanding; rest already merged.

msbutler commented on pkg/sql/crdb_internal.go:
```diff
@@ -1179,32 +1179,55 @@ CREATE TABLE crdb_internal.jobs (
+	settings.ApplicationLevel,
+	"sql.jobs.legacy_vtable.enabled",
+	"cause the crdb_internal.jobs vtable to be produced from the legacy payload info records",
+	false, // TODO(dt): flip this once we add permissive auth checks.
```
can we make this metamorphic to maintain test coverage of the old vtable? Also i think the todo can be removed.

dt replied:
We get it in mixed version. I was planning     on removing the setting outright as soon as the backport lands so there won’t be any coverage to maintain.

msbutler replied:
I think we want to maintain coverage on the 25.1 release branch?

dt replied:
we have that via the version gate already though. This isn't a public setting so I'm not sure it seems worth an extra metamorphic just to cover it?

msbutler replied:
resolved in person. bors away!

dt commented:
TFTR!

bors r+
